File: app/src/main/java/com/openclassrooms/realestatemanager/ViewModelFactory.java
package com.openclassrooms.realestatemanager;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

public class ViewModelFactory implements ViewModelProvider.Factory {

    private static ViewModelFactory sFactory;

    private final RealEstateRepository realEstateRepository;
    private final PhotoRepository photoRepository;

    public ViewModelFactory(
            RealEstateRepository realEstateRepository,
            PhotoRepository photoRepository
    ) {
        this.realEstateRepository = realEstateRepository;
        this.photoRepository = photoRepository;
    }

    /*
    public static ViewModelFactory getInstance() {
        if (sFactory == null) {
            synchronized (ViewModelFactory.class) {
                if (sFactory == null) {
                    sFactory = new ViewModelFactory(
                            new RealEstateRepository()
                    );
                }
            }
        }

        return sFactory;
    }

     */

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        if(modelClass.isAssignableFrom(SharedViewModel.class)){
            return (T) new SharedViewModel(realEstateRepository, photoRepository);
        }
        throw new IllegalArgumentException("Unknown ViewModel class");
    }
}

File: app/src/main/java/com/openclassrooms/realestatemanager/injection/ViewModelFactory.kt
package com.openclassrooms.realestatemanager.injection

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.openclassrooms.realestatemanager.repository.PhotoRepository
import com.openclassrooms.realestatemanager.repository.RealEstateRepository
import com.openclassrooms.realestatemanager.ui.edit.RealEstateAddViewModel

class ViewModelFactory(
        private val realEstateRepository: RealEstateRepository,
        private val photoRepository: PhotoRepository
) : ViewModelProvider.Factory {

    override fun <T : ViewModel?> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(RealEstateAddViewModel::class.java)) {
            @Suppress("UNCHECKED_CAST")
            return RealEstateAddViewModel(realEstateRepository) as T
        }

        throw IllegalArgumentException("Unknown ViewModel class")
    }
}